File: forms/cardiopulmonary-resuscitation-training/full-stack-with-loco-tera-htmx-alpine/src/engine/bls_grader.rs
use super::bls_rules::all_rules;
use super::flagged_issues::detect_additional_flags;
use super::types::{AssessmentData, FiredRule, GradingResult, Outcome};
use super::utils::NON_CRITICAL_DEFICIENCY_LIMIT;

/// Pure function: grades a BLS Skills Verification case per AHA pass/fail logic.
///
/// - Any critical-action rule firing `'no'` -> overall Fail.
/// - More than two non-critical `'no'` deficiencies -> Fail.
/// - Nothing assessed yet -> Fail (so examiner is prompted to actually mark items).
/// - Otherwise -> Pass.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let rules = all_rules();
    let total_rules = rules.len() as u32;

    let mut fired_rules: Vec<FiredRule> = Vec::with_capacity(rules.len());
    let mut critical_failures: Vec<FiredRule> = Vec::new();
    let mut non_critical_deficiencies: Vec<FiredRule> = Vec::new();
    let mut answered_count: u32 = 0;

    for rule in rules {
        let status = (rule.evaluate)(data);
        let entry = FiredRule {
            id: rule.id.to_string(),
            step: rule.step,
            category: rule.category.to_string(),
            description: rule.label.to_string(),
            critical: rule.critical,
            status: status.clone(),
        };
        if status == "yes" || status == "no" {
            answered_count += 1;
        }
        if status == "no" {
            if rule.critical {
                critical_failures.push(entry.clone());
            } else {
                non_critical_deficiencies.push(entry.clone());
            }
        }
        fired_rules.push(entry);
    }

    let outcome: Outcome = if !critical_failures.is_empty() {
        "fail".to_string()
    } else if non_critical_deficiencies.len() > NON_CRITICAL_DEFICIENCY_LIMIT {
        "fail".to_string()
    } else if answered_count == 0 {
        "fail".to_string()
    } else {
        "pass".to_string()
    };

    let additional_flags = detect_additional_flags(data, &critical_failures, &non_critical_deficiencies);

    GradingResult {
        outcome,
        critical_failures,
        non_critical_deficiencies,
        fired_rules,
        additional_flags,
        answered_count,
        total_rules,
        timestamp,
    }
}

/// Run the rules without classification — handy for testing.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out = Vec::new();
    for rule in all_rules() {
        let status = (rule.evaluate)(data);
        out.push(FiredRule {
            id: rule.id.to_string(),
            step: rule.step,
            category: rule.category.to_string(),
            description: rule.label.to_string(),
            critical: rule.critical,
            status,
        });
    }
    out
}

File: forms/cardiopulmonary-resuscitation-training/front-end-with-svelte/src/lib/engine/bls-grader.ts
import type { AssessmentData, Outcome, FiredRule, TriState, GradingResult } from './types';
import { blsRules } from './bls-rules';
import { detectAdditionalFlags } from './flagged-issues';

/**
 * Maximum number of non-critical deficiencies tolerated before the overall
 * outcome is a Fail.
 */
export const NON_CRITICAL_DEFICIENCY_LIMIT = 2;

/**
 * Pure function: evaluates every BLS rule in the registry against the trainee's
 * observed performance and classifies the overall pass/fail outcome.
 *
 * AHA pass/fail logic:
 *   - Any critical-action rule firing 'no' -> overall Fail.
 *   - More than two non-critical 'no' deficiencies -> Fail.
 *   - Nothing assessed yet -> Fail (prompt the examiner to record items).
 *   - Otherwise -> Pass.
 *   - Items the examiner has not assessed ('na' or '') do not count as
 *     deficiencies.
 */
export function gradeBLS(data: AssessmentData): GradingResult {
	const firedRules: FiredRule[] = [];
	const criticalFailures: FiredRule[] = [];
	const nonCriticalDeficiencies: FiredRule[] = [];
	let answeredCount = 0;

	for (const rule of blsRules) {
		let status: TriState = '';
		try {
			status = rule.evaluate(data);
		} catch (e) {
			console.warn(`BLS rule ${rule.id} evaluation failed:`, e);
			status = '';
		}

		const entry: FiredRule = {
			id: rule.id,
			step: rule.step,
			category: rule.category,
			description: rule.label,
			critical: rule.critical,
			status
		};
		firedRules.push(entry);

		if (status === 'yes' || status === 'no') answeredCount++;

		if (status === 'no') {
			if (rule.critical) {
				criticalFailures.push(entry);
			} else {
				nonCriticalDeficiencies.push(entry);
			}
		}
	}

	let outcome: Outcome;
	if (criticalFailures.length > 0) {
		outcome = 'fail';
	} else if (nonCriticalDeficiencies.length > NON_CRITICAL_DEFICIENCY_LIMIT) {
		outcome = 'fail';
	} else if (answeredCount === 0) {
		// Nothing assessed yet — surface a neutral "fail" on submission so the
		// examiner is prompted to actually mark items, but don't claim success.
		outcome = 'fail';
	} else {
		outcome = 'pass';
	}

	const additionalFlags = detectAdditionalFlags(data, {
		criticalFailures,
		nonCriticalDeficiencies
	});

	return {
		outcome,
		criticalFailures,
		nonCriticalDeficiencies,
		firedRules,
		additionalFlags,
		answeredCount,
		totalRules: blsRules.length,
		timestamp: new Date().toISOString()
	};
}
